pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -241,7 +241,8 @@ func (s *Scheduler) add(ctx context.Context, pod *v1.Pod) error {
 			instanceTypes = filterByRemainingResources(s.instanceTypes[machineTemplate.ProvisionerName], remaining)
 			if len(instanceTypes) == 0 {
 				errs = multierr.Append(errs, fmt.Errorf("all available instance types exceed limits for provisioner: %q", machineTemplate.ProvisionerName))
-				continue
+				// Return so we don't pick the same instance from a different provisioner
+				return errs
```

Comment:
Falling back across provisioners with limits has some valid use cases.  One is where you use a high priority provisioner with limits to ensure that you use all of your reserved instances first. Once those are gone (the limit is met), you fall back to a lower weighted provisioner and launch whatever is cheapest.

Reply:
is this the wrong place to address https://github.com/aws/karpenter/issues/3106 ?

The issue states "near" limits, and the above seems to be in the flow where you've literally reached the limit.

So I'm probably putting this return in the wrong place.

cc @jonathan-innis

Reply:
The use-case for Reserved Instances could be that the limit hasn't quite been breached yet, but there isn't sufficient resource space in the highest priority provisioner to create a machine in the worst case (biggest) of the instance type options on that machine. 

For example, if our instance type options were something like 2cpus, 4cpus, 8cpus on a machine, Karpenter would calculate the limit predicting 8cpus, even though we're likely to get a 2 cpu vm so that we never overshoot the limit.  If we're scheduling more than one machine, we'd fallback to the next provisioner based on the 8cpus, but when 2 cpus is actually launched, we actually had 6 cpus to spare. 

We previously did overshoot the limit and that was surprising to a lot of users since a limit... wasn't really a limit. 

I'm not sure about the solution that @jonathan-innis proposed since I think that takes us back to overshooting the limit which I'm not sure we want to do.  Maybe there's a heuristic we could use like the median CPUs of the options, but spot VMs are more likely to result in overshooting the limit since we can't easily predict what instance type is going to be launched.

Reply:
We need a way to dilineate the cases. We basically want the following behavior:

1. Don't fallback to a lower-weighted provisioner when a higher-weighted provisioner _could_ schedule the capacity
2. Fallback to a lower-weighted provisioner when a higher-weighted provisioner has no way to schedule the capacity

Is there a way that we can describe an exit-early strategy based purely off of breaching provisioner limits combined with already having scheduled capacity to the provisioner in this scheduling round? i.e. we basically need to know if there is "hypothetical capacity" assigned to a provisioner that hasn't been resolved yet that might be causing us to think that we could breach our limits. That's the crux of the problem here that's causing us to fallback early.